### PR TITLE
[Sema] Setter has incorrect mutating-ness inside class-constrained protocol extension

### PR DESCRIPTION
**Pull request by theblixguy**
If we have a class-constrained protocol extension, where the protocol does not impose a class requirement, then the setter is incorrectly inferred to be mutating by default.

Resolves [SR-11298](https://bugs.swift.org/browse/SR-11298).

### CODE REVIEW

**Comment by theblixguy** [comment]
cc @slavapestov

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -1995,6 +1995,13 @@ SelfAccessKindRequest::evaluate(Evaluator &evaluator, FuncDecl *FD) const {
...
     case AccessorKind::Modify:
+      // If we have a class-constrained protocol extension, where the
+      // protocol does not have a class requirement, then the setter is
+      // implicitly nonmutating, unless it has been explicitly marked as
+      // mutating.
+      if (AD->isEligibleToBeNonMutatingInsideClassConstrainedProtoExt())
+        return SelfAccessKind::NonMutating;
+
       if (AD->isInstanceMember() &&
           doesContextHaveValueSemantics(AD->getDeclContext()))
```
Comment (slavapestov) [inline lib/Sema/TypeCheckDecl.cpp]: Can you investigate why doesContextHaveValueSemantics() return true for a class-constrained extension? Can we fix the problem there instead?

Comment (theblixguy) [reply lib/Sema/TypeCheckDecl.cpp]: @slavapestov Basically, `doesContextHaveValueSemantics()` gets the DeclContext's declared interface type, which in this case is the extended protocol's type, which doesn't impose a class constraint so the method returns true.

Comment (theblixguy) [reply lib/Sema/TypeCheckDecl.cpp]: I think what we could do is to add another check to `doesContextHaveValueSemantics` where if the decl context is an extension then we check whether the extension's generic signature requires Self to be a class (i.e. the same check I added in this patch, but moved into `doesContextHaveValueSemantics()` instead).

**Comment by jrose-apple** [comment]
Technically source-breaking, but I kinda hope no one is depending on this.

```swift
protocol SR_11298_P {
  init()
}

class SR_11298_C: SR_11298_P {
  var property: String = ""
  required init() {}
}

extension SR_11298_P where Self: SR_11298_C {
  var wrappingProperty: String {
    get { return property }
    set { self = .init() }
  }
}

let origInstance = SR_11298_C()
var instance = origInstance
instance.wrappingProperty = ""
print(instance === origInstance)
```

**Review by slavapestov — APPROVED** [review]

**Comment by theblixguy** [comment]
Can someone kick off the tests? Thanks!

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by theblixguy** [comment]
Thanks! All tests have passed!

**Comment by jrose-apple** [comment]
Thanks, @theblixguy!

**Comment by hamishknight** [comment]
Should we make a note of this in the changelog? Note that this change can also break code where the setter delegates to another mutating requirement or extension member, such as:

```swift
protocol P {
  var property: String { get set }
}

extension P where Self : AnyObject {
  var wrappingProperty: String {
    get { property }
    set { property = newValue }
  }
}
```

**Comment by jrose-apple** [comment]
Ah, hm. That might be a more serious concern, especially since people are already confused about why protocol methods can be `mutating` for classes.

**Comment by jrose-apple** [comment]
Reverting for now.

**Comment by slavapestov** [comment]
@theblixguy Can you create a PR that adds a test case for the current behavior, including @hamishknight's example? I'm not sure it's tested anywhere in our suite and we want to make sure any change is intentional.

**Comment by theblixguy** [comment]
Sure. I’ve also created a [forum thread](https://forums.swift.org/t/class-constrained-protocol-extension-has-incorrect-mutatingness/27962) to discuss this.

**Pull request merged**